gensim/models/utils_any2vec.py

```diff
@@ -88,44 +102,80 @@ def _ft_hash_py_broken(string):
 
 try:
     from gensim.models._utils_any2vec import (
-        ft_hash as _ft_hash_cy,
+        compute_ngrams,
         ft_hash_broken as _ft_hash_cy_broken,
-        compute_ngrams as _compute_ngrams
+        ft_hash_bytes as _ft_hash_cy_bytes,
     )
-    _ft_hash = _ft_hash_cy
+    _ft_hash_bytes = _ft_hash_cy_bytes
     _ft_hash_broken = _ft_hash_cy_broken
+    FAST_VERSION = 0
 except ImportError:
+    # failed... fall back to plain python
     FAST_VERSION = -1
-
-    _ft_hash = _ft_hash_py
+    _ft_hash_bytes = _ft_hash_py_bytes
     _ft_hash_broken = _ft_hash_py_broken
 
-    # failed... fall back to plain python
-    def _compute_ngrams(word, min_n, max_n):
-        """Get the list of all possible ngrams for a given word.
 
-        Parameters
-        ----------
-        word : str
-            The word whose ngrams need to be computed.
-        min_n : int
-            Minimum character length of the ngrams.
-        max_n : int
-            Maximum character length of the ngrams.
-
-        Returns
-        -------
-        list of str
-            Sequence of character ngrams.
+def ft_ngram_hashes(word, minn, maxn, num_buckets):
+    """Calculate the ngrams of the word and hash them.
 
-        """
-        BOW, EOW = ('<', '>')  # Used by FastText to attach to all words as prefix and suffix
-        extended_word = BOW + word + EOW
-        ngrams = []
-        for ngram_length in range(min_n, min(len(extended_word), max_n) + 1):
-            for i in range(0, len(extended_word) - ngram_length + 1):
-                ngrams.append(extended_word[i:i + ngram_length])
-        return ngrams
+    Do this in a way that is compatible with the original Facebook implementation.
+
+    Parameters
+    ----------
+    word : str
+        The word to calculate ngram hashes for.
+    minn : int
+        Minimum ngram length
+    maxn : int
+        Maximum ngram length
+    num_buckets : int
+        The number of buckets
+
+    Returns
+    -------
+        A list of hashes (integers), one per each detected ngram.
+
+    """
+    ngrams = compute_ngrams(word, minn, maxn)
+    #
+    # This is a trick to avoid encoding each ngram separately, which is
+    # computationally expensive.  It works for two reasons:
+    #
+    # 1) ngrams are guaranteed to _not_ have any spaces, because that is the
+    # character we use to split sentences into words.
+    # 2) space is an ASCII character, so it survives UTF-8 encoding unchanged.
+    #
+    encoded_ngrams = (" ".join(ngrams)).encode("utf-8").split()
```

Comment:
I'm not sure what's better:
- such one liner to encode in bytes or
- just call encode for each str in `_ft_hash`

Reply:
The benchmarks suggest that the former is better, don't they?

https://github.com/RaRe-Technologies/gensim/issues/1261#issuecomment-455790163

Reply:
yes, they do. I just wonder why it's there.

Reply:
Why do you think encoding is not a bottleneck? In my opinion, that's exactly where the problem is.

The new version does one join, one encode, and one split per word.

The older version did an encode per ngram. There can be many more ngrams than words, especially for longer words. Furthermore, the ngrams overlap, meaning you're unnecessarily encoding the same character multiple times. This optimization wasn't my idea, but my limited understanding suggests that avoiding the unnecessary encode calls is where the win comes from.

@menshikh-iv Can you please comment?

Reply:
1. This line isn't a bottleneck according to profiling result
2. I'm only worried about "terminator" here (not sure, but some wrong stuff happens and `" "` potentially can be part of token (though shouldn't be), for this reason, I guess, better to generate ngrams from bytes (same as FB) + use hashing in same manner (if ngram found - calc hash now, don't collect all of this)

Reply:
OK, I copied your port of the FB implementation.